### PR TITLE
[Enhancement] Support concat function for array type

### LINKED ISSUE #22201: Support concat function with arguments are array type
## Enhancement
#14825
StarRocks do not support concat function which arguments are array type, user need to use array_concat to wrok

### PR DESCRIPTION
**Pull request by Youngwb**
## What type of PR is this：
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #22201

## Problem Summary(Required) ：

For trino parser on starrocks, function conversion can be performed at the syntax level, However, for some conversions, it cannot convert when need to know the argument type, such like `select concat(c1, c2)`, parser can not get the type of c1 and c2, it can not decide to use `concat` or `array_concat` in SR.
So, we need to do it in SR anlyze stage:
when **concat** function arguments have array type, use **array_concat** to replace it

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto backported to target branch
  - [x] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 15 / 15 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/ExpressionAnalyzer.java|15|15|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 17 / 17 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/ExpressionAnalyzer.java|17|17|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/ExpressionAnalyzer.java [hunk 1]
```diff
@@ -974,7 +1000,7 @@ public Void visitFunctionCall(FunctionCallExpr node, Scope scope) {
             } else if (FunctionSet.ARRAY_GENERATE.equals(fnName)) {
                 fn = getArrayGenerateFunction(node);
                 argumentTypes = node.getChildren().stream().map(Expr::getType).toArray(Type[]::new);
-            } else {
+            }  else {
```
Comment (DorianZheng) [inline fe/fe-core/src/main/java/com/starrocks/sql/analyzer/ExpressionAnalyzer.java:1003]: duplicate empty space

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 17 / 17 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/ExpressionAnalyzer.java|17|17|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/connector/parser/trino/Trino2SRFunctionCallTransformer.java [hunk 2]
```diff
@@ -109,10 +108,6 @@ private static void registerArrayFunctionTransformer() {
                         new PlaceholderExpr(1, Expr.class), new PlaceholderExpr(2, Expr.class)
                 )))));
 
-        // 2. concat -> array_concat
```
Comment (DorianZheng) [inline fe/fe-core/src/main/java/com/starrocks/connector/parser/trino/Trino2SRFunctionCallTransformer.java:112]: could you add more description for this PR, why previous method won't work?

Comment (Youngwb) [reply fe/fe-core/src/main/java/com/starrocks/connector/parser/trino/Trino2SRFunctionCallTransformer.java:112]: Done

**Review by kangkaisen — APPROVED** [review]

**Review by Astralidea — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.0

**Pull request merged**